library/init/meta/interactive.lean

```diff
@@ -6,8 +6,11 @@ Authors: Leonardo de Moura
 prelude
 import init.meta.tactic init.meta.rewrite_tactic init.meta.simp_tactic
 import init.meta.smt.congruence_closure init.category.combinators
+import init.meta.lean_parser
 
 namespace interactive
+@[reducible] meta def parse {α : Type} (p : lean_parser α) : Type := α
```

Comment:
> @leodemoura Can't we make the error message much better by checking the type of the returned pre-expression? That is, when we evaluate the parser, we already know the type α. We can then elaborate the returned pre-expression, and check whether it has indeed type α.

At parsing time, we can elaborate closed terms. However, anything that contains open terms may fail to elaborate since it depends on the local context which has not been elaborated yet. For example, consider
```lean
meta def min_tac (a b : interactive.types.qexpr) : tactic unit :=
let x := `(%%a + 1),
     y := `(%%b + 1), 
in `[apply @by_cases (%%x ≤ %%y), repeat {min_tac_step}]
```
The subterm `@by_cases (%%x ≤ %%y)` produces the open `pexpr`
```lean
(pexpr.subst (pexpr.subst `(λ (_x_1 : _) (_x_2 : _), @by_cases (le _x_1 _x_2)) (to_pexpr x)) (to_pexpr y))
```
We cannot elaborate it at parsing time since we don't know the types of `x` and `y`.